integration_tests/lock_test.go

```diff
@@ -1591,3 +1591,142 @@ func (s *testLockWithTiKVSuite) TestPessimisticRollbackWithRead() {
+	// max_execution_time (200ms) < lock wait time (800ms) < TiKV limit (1000ms)
+	txn3, err := s.store.Begin()
+	s.NoError(err)
+	txn3.SetPessimistic(true)
+
+	startTime := time.Now()
+	lockCtx3 := kv.NewLockCtx(txn3.StartTS(), 800, startTime)             // 800ms lock wait
+	lockCtx3.MaxExecutionDeadline = startTime.Add(200 * time.Millisecond) // But max exec time only 200ms
+
+	err = txn3.LockKeys(ctx, lockCtx3, k1)
+	elapsed := time.Since(startTime)
+
+	s.Error(err)
+	s.ErrorAs(err, &queryInterruptedErr)
+	s.Equal(uint32(transaction.MaxExecTimeExceededSignal), queryInterruptedErr.Signal)
+
+	// Should timeout around 200ms, not 800ms
+	s.Greater(elapsed, 190*time.Millisecond)
+	s.Less(elapsed, 400*time.Millisecond)
+
+	// Test case 3: lock wait timeout shorter than max_execution_time
+	// lock wait time (150ms) < max_execution_time (600ms) < TiKV limit (1000ms)
+	txn4, err := s.store.Begin()
+	s.NoError(err)
+	txn4.SetPessimistic(true)
+
+	startTime = time.Now()
+	lockCtx4 := kv.NewLockCtx(txn4.StartTS(), 150, startTime)             // 150ms lock wait
+	lockCtx4.MaxExecutionDeadline = startTime.Add(600 * time.Millisecond) // 600ms max exec time
+
+	err = txn4.LockKeys(ctx, lockCtx4, k1)
+	elapsed = time.Since(startTime)
+
+	s.Error(err)
+	// Should be lock wait timeout, not max execution time error
+	s.Equal(tikverr.ErrLockWaitTimeout.Error(), err.Error())
+
+	// Should timeout around 150ms
+	s.Greater(elapsed, 120*time.Millisecond)
+	s.Less(elapsed, 300*time.Millisecond)
+
+	// Test case 4: TiKV limit is the constraint
+	// max_execution_time (1200ms) > TiKV limit (1000ms) > lock wait time (900ms)
+	txn5, err := s.store.Begin()
+	s.NoError(err)
+	txn5.SetPessimistic(true)
+
+	startTime = time.Now()
+	lockCtx5 := kv.NewLockCtx(txn5.StartTS(), 900, startTime)              // 900ms lock wait
+	lockCtx5.MaxExecutionDeadline = startTime.Add(1200 * time.Millisecond) // 1200ms max exec time
+
+	err = txn5.LockKeys(ctx, lockCtx5, k1)
+	elapsed = time.Since(startTime)
+
+	s.Error(err)
+	// Should be lock wait timeout due to TiKV's 1-second cap, not max execution time
+	s.Equal(tikverr.ErrLockWaitTimeout.Error(), err.Error())
+
+	// Should timeout around 900ms (limited by requested lock wait time)
+	s.Greater(elapsed, 800*time.Millisecond)
+	s.Less(elapsed, 1200*time.Millisecond)
+
+	// Test case 5: No max_execution_time deadline (normal behavior)
+	txn6, err := s.store.Begin()
+	s.NoError(err)
+	txn6.SetPessimistic(true)
+
+	lockCtx6 := kv.NewLockCtx(txn6.StartTS(), kv.LockNoWait, time.Now())
+	// MaxExecutionDeadline defaults to zero time (no deadline)
+
+	err = txn6.LockKeys(ctx, lockCtx6, k1)
+	s.Error(err)
+	// Should be immediate lock acquisition failure, not max execution time error
+	s.Equal(tikverr.ErrLockAcquireFailAndNoWaitSet.Error(), err.Error())
+
+	// Cleanup: Unlock k1 and test successful lock with max_execution_time
+	s.NoError(txn1.Rollback())
+
+	// Test case 6: Successful lock acquisition with max_execution_time set
```

Comment:
Is there a test case to verify after timeout the resolve lock should not be triggered?